classloader-leak-prevention/classloader-leak-prevention-core/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/DriverManagerCleanUp.java

```diff
@@ -5,64 +5,56 @@
-        }
-
-        for (Driver driver : driversToDeregister) {
-            try {
-                preventor.warn("JDBC driver loaded by protected ClassLoader deregistered: " + driver.getClass());
-                DriverManager.deregisterDriver(driver);
-            }
-            catch (SQLException e) {
-                preventor.error(e);
-            }
-        }
-    }
-
-    /**
-     * add getAllDrivers method. DriverManager.getDrivers() only return the drivers which be load by caller(DriverManagerCleanUp.class).
-     * For many scenarios the Caller is not the same classloader which load the jdbc Drivers
-     *
-     * @param preventor
-     * @return
-     */
-    public java.util.Enumeration<Driver> getAllDrivers(ClassLoaderLeakPreventor preventor) {
-        java.util.Vector<Driver> result = new java.util.Vector<Driver>();
+  @Override
+  public void cleanUp(ClassLoaderLeakPreventor preventor) {
+    final Enumeration<Driver> allDrivers = getAllDrivers(preventor);
+    while (allDrivers.hasMoreElements()) {
+      final Driver driver = allDrivers.nextElement();
+      if (preventor.isLoadedInClassLoader(driver)) {
         try {
-            CopyOnWriteArrayList<?> driverinfos = preventor.getStaticFieldValue(DriverManager.class, "registeredDrivers");
-            for (Object driverinfo : driverinfos) {
-                Driver driver = (Driver) preventor.getFieldValue(driverinfo, "driver");
-                result.addElement(driver);
-            }
-        }
-        catch (Exception e) {
-            preventor.warn("get All registeredDrivers Exception ");
-            preventor.error(e);
-
+          preventor.warn("JDBC driver loaded by protected ClassLoader deregistered: " + driver.getClass());
+          DriverManager.deregisterDriver(driver);
+        } catch (SQLException e) {
+          preventor.error(e);
         }
-        return (result.elements());
+      }
+    }
+  }
 
+  /**
+   * Add getAllDrivers method. DriverManager.getDrivers() only return the drivers
+   * which be load by caller(DriverManagerCleanUp.class). For many scenarios the
+   * caller is not the same classloader which load the jdbc drivers.
+   * @param preventor
+   * @return All drivers in DriverManager's registeredDrivers field,or
+   *         DriverManager.getDrivers() if exception occurred
+   */
+  public Enumeration<Driver> getAllDrivers(ClassLoaderLeakPreventor preventor) {
+    Vector<Driver> result = new java.util.Vector<Driver>();
+    try {
+      ArrayList<?> driverinfos = preventor.getStaticFieldValue(DriverManager.class, "registeredDrivers");
+      for (Object driverinfo : driverinfos) {
+        Driver driver = (Driver) preventor.getFieldValue(driverinfo, "driver");
+        if (driver == null)
+          throw new NullPointerException();
+        else
+          result.addElement(driver);
+      }
+    } catch (Exception e) {
+      preventor.warn("get All registeredDrivers Exception");
+      return DriverManager.getDrivers();

```

Comment:
The code now assumes the returned Enumeration does not cause `ConcurrentModificationException` if driver is removed while looping. Is that guaranteed by the API contract, or should we make of copy to ensure the library does not start failing in case the `DriverManager` internals happens to be changed?

Reply:
I think it's no reason the jdk developer to changed the default manner of getDrivers. they are already considered  avoid ConcurrentModificationException in current code.